cryodrgn/fft.py

```diff
@@ -1,44 +1,108 @@
 
-    return pp.fft.fftshift(
-        pp.fft.fft2(pp.fft.fftshift(img, axes=(-1, -2))), axes=(-1, -2)
+def transform_in_chunks(
+    f, data, chunksize=None, n_workers=1, inplace=False, *args, **kwargs
+):
+    def g(src, dest, indices):
+        logger.debug(f"Transforming chunk of size {len(indices)}")
+        dest[indices, ...] = f(src[indices, ...], *args, **kwargs)
+
+    ndim = data.ndim
+    if data.ndim == 2:
+        n = 1
+        src = data[np.newaxis, ...]
+    elif data.ndim == 3:
+        n = data.shape[0]
+        src = data
+    else:
+        raise RuntimeError(f"Unsupported data ndim {ndim}")
+
+    indices = np.arange(n)
+    chunksize = min(n, chunksize) if chunksize is not None else n
+    chunk_indices = np.array_split(indices, n // chunksize)
+    n_chunks = len(chunk_indices)
+
+    if inplace:
+        dest = src
+    else:
+        if isinstance(src, np.ndarray):
+            dest = src.copy()
+        elif isinstance(src, torch.Tensor):
+            dest = src.detach().clone()
+        else:
+            raise RuntimeError("Unsupported type for src")
+
+    with Pool(n_workers) as pool:
+        pool.starmap(
+            g, zip(repeat(src, n_chunks), repeat(dest, n_chunks), chunk_indices)
+        )
+        return dest.reshape(data.shape)
+
+
+def normalize(
+    img, mean=0, std=None, std_n=None, inplace=False, chunksize=None, n_workers=1
+):
+    def _normalize(img, mean, std):
+        return (img - mean) / std
+
+    if std is None:
+        # Since std is a memory consuming process, use the first std_n samples for std determination
+        std = torch.std(img[:std_n, ...])
+
+    retVal = transform_in_chunks(
+        _normalize,
+        img,
+        chunksize=chunksize,
+        n_workers=n_workers,
+        inplace=inplace,
+        mean=mean,
+        std=std,
     )
 
+    logger.info(f"Normalized by {mean} +/- {std}")
+    return retVal
 
-def fftn_center(img):
-    pp = np if isinstance(img, np.ndarray) else cp
 
-    return pp.fft.fftshift(pp.fft.fftn(pp.fft.fftshift(img)))
+def fft2_center(img):
+    return fftshift(fft2(fftshift(img, dim=(-1, -2))), dim=(-1, -2))
+
 
+def fftn_center(img):
+    return fftshift(fftn(fftshift(img)))
 
-def ifftn_center(V):
-    pp = np if isinstance(V, np.ndarray) else cp
 
-    V = pp.fft.ifftshift(V)
-    V = pp.fft.ifftn(V)
-    V = pp.fft.ifftshift(V)
-    return V
+def ifftn_center(img):
+    if isinstance(img, np.ndarray):
+        # Note: We can't just typecast a complex ndarray using torch.Tensor(array) !
+        img = torch.complex(torch.Tensor(img.real), torch.Tensor(img.imag))
+    x = ifftshift(img)
+    y = ifftn(x)
+    z = ifftshift(y)
+    return z
 
 
-def ht2_center(img):
-    f = fft2_center(img)
-    return f.real - f.imag
+def ht2_center(img, inplace=False, chunksize=None, n_workers=1):
+    def _ht2_center(img):
+        _img = fft2_center(img)
+        return _img.real - _img.imag
 
+    return transform_in_chunks(
```

Comment:
You should be calling these functions on the GPU, so no need to parallelize (and indeed, parallelizing at the Python level will probably lead to issues because it will spawn processes that will need to create their own cuda contexts, etc. Very bad.)

Reply:
Great - will go away.

Reply:
Removed in commit 7f3b87558d77091d39877540c5a6906de0ea41cc